src/app/records/Record.tsx

```diff
@@ -276,6 +276,28 @@ export function RecordGrid({
   // TODO: was there really a problem with duplicate records?
   const seenRecords = new Set<number>();
 
+  // I need to do this check so it wont have an error about recordTypeName not existing
+  if (
+    'recordTypeName' in records[0].recordDef &&
+    records[0].recordDef.recordTypeName === 'Weapon Pattern'
+  ) {
```

Comment:
Unfortunately `recordTypeName` is localized by Bungie, so this won't work for users who use DIM in a different language.

Also this component is responsible for showing tracked triumphs too, so this will re-order them by completion if the user happens to have a weapon pattern first in the list of tracked triumphs.

Maybe this is better addressed in the part of the code that creates the DimRecords, and weapon patterns identified by grandparent presentation node hash?

Reply:
I believe the spirit of the original request was to have a "sort by" option for records (like we do for loadouts) that allows all records, not just weapon patterns, to be re-sorted by completion percentage.

Reply:
I'm having trouble figuring out how I would need to check if the what I have is a weapon pattern. Right now I'm trying to see if there is a way to match the hash with a hash in the craftable-hashes.json using getCraftingTemplate but I don't think this is the right way to do it.  Also, you are saying that this should be done in PresentationNodeLeaf?

I only did the weapon pattern sorting because I did not want try and write code for the entire records page because I knew I would probably not do a lot of it correctly. I wanted to to it segment by segment. I'm assuming the end product would be a toggle button under "Reveal redacted triumphs" that will sort the entire Records page right?

Reply:
Right, a single toggle with no special per-section logic, which also means no need to solve the problem of identifying particular sections.

Reply:
You would still need to specify different sections though since some of them use different wording for tracking their progression. Weapon patterns uses `recordComponent.objectives[0].progress` while triumphs use `recordComponent.intervalObjectives[2].progress`. 
![Capture](https://github.com/DestinyItemManager/DIM/assets/99226958/1ed2f4fc-c640-4fc1-b291-56eb70531588)

For exotic weapons and armor in collections, I would need to check to check if `owned`  is true or not to push it to the top of the list though for some reason they are all false. 

I suppose I could start buy only sorting the tabs themselves since they all seem to use `visible` and `acquired`  but if I want to sort each item in each RecordGrid it seems like they would need to have unique logic.

Reply:
`intervalObjectives` are simply a different kind of measuring progress. Maybe it could work like this: If a record has `intervalObjectives`, use those to calculate completion percentage, otherwise use the arithmetic average of the individual `objectives` completion percentage or something.

Reply:
Will this line work? `(entry) => entry['acquired'] / entry['visible'],` I am not sure if this will have issues with language.